config/config.go

```diff
@@ -0,0 +1,229 @@
+// limitations under the License.
+
+package config
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"gopkg.in/ini.v1"
+)
+
+var (
+	configReloadSuccess = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "mysqld_exporter",
+		Name:      "config_last_reload_successful",
+		Help:      "Mysqld exporter config loaded successfully.",
+	})
+
+	configReloadSeconds = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "mysqld_exporter",
+		Name:      "config_last_reload_success_timestamp_seconds",
+		Help:      "Timestamp of the last successful configuration reload.",
+	})
+
+	cfg *ini.File
+
+	opts = ini.LoadOptions{
+		// Do not error on nonexistent file to allow empty string as filename input
+		Loose: true,
+		// MySQL ini file can have boolean keys.
+		AllowBooleanKeys: true,
+	}
+
+	err error
+)
+
+type Config struct {
+	Sections map[string]MySqlConfig
+}
+
+type MySqlConfig struct {
+	User                  string `ini:"user"`
+	Password              string `ini:"password"`
+	Host                  string `ini:"host"`
+	Port                  int    `ini:"port"`
+	Socket                string `ini:"socket"`
+	SslCa                 string `ini:"ssl-ca"`
+	SslCert               string `ini:"ssl-cert"`
+	SslKey                string `ini:"ssl-key"`
+	TlsInsecureSkipVerify bool   `ini:"ssl-skip-verfication"`
+}
+
+type MySqlConfigHandler struct {
+	sync.RWMutex
+	TlsInsecureSkipVerify bool
+	Config                *Config
+}
+
+func (ch *MySqlConfigHandler) GetConfig() *Config {
+	ch.RLock()
+	defer ch.RUnlock()
+	return ch.Config
+}
+
+func (ch *MySqlConfigHandler) ReloadConfig(filename string, mysqldAddress string, mysqldUser string, tlsInsecureSkipVerify bool, logger log.Logger) error {
+	var host, port string
+	defer func() {
+		if err != nil {
+			configReloadSuccess.Set(0)
+		} else {
+			configReloadSuccess.Set(1)
+			configReloadSeconds.SetToCurrentTime()
+		}
+	}()
+
+	if cfg, err = ini.LoadSources(
+		opts,
+		[]byte("[client]\npassword = ${MYSQLD_EXPORTER_PASSWORD}\n"),
+		filename,
+	); err != nil {
+		return fmt.Errorf("failed to load %s", filename)
+	}
+
+	if host, port, err = net.SplitHostPort(mysqldAddress); err != nil {
+		return fmt.Errorf("failed to parse address: %s", err)
+	}
+
+	if clientSection := cfg.Section("client"); clientSection != nil {
+		if cfgHost := clientSection.Key("host"); cfgHost.String() == "" {
+			cfgHost.SetValue(host)
+		}
+		if cfgPort := clientSection.Key("port"); cfgPort.String() == "" {
+			cfgPort.SetValue(port)
+		}
+		if cfgUser := clientSection.Key("user"); cfgUser.String() == "" {
+			cfgUser.SetValue(mysqldUser)
+		}
+	}
+
+	cfg.ValueMapper = os.ExpandEnv
+	config := &Config{}
+	m := make(map[string]MySqlConfig)
+	for _, sec := range cfg.Sections() {
+		sectionName := sec.Name()
+
+		if sectionName == "DEFAULT" {
```

Comment:
we should skip client section as well

Reply:
That would break configuration for those that are currently using it with a config file rather than `DATA_SOURCE_NAME`, are we OK with that?

Reply:
No, I think we need to keep this so it continues to work as-is.